connectors/netty-connector/src/main/java/org/glassfish/jersey/netty/connector/internal/NettyInputStream.java

```diff
@@ -16,107 +16,150 @@
-        }
+             }
+
+             throw new IOException(cause);
+          }
+       }
+
+       int rem = buffer.remaining();
+       if (rem < len) {
+          len = rem;
+       }
+       buffer.get(b, off, len);
+       if (rem == len) {
+          releaseByteBuf();
+       }
+
+       return len;
     }
 
     @Override
     public int read() throws IOException {
+       if (ONE_BYTE == null) {
+          ONE_BYTE = new byte[1];
+       }
+       int r = read(ONE_BYTE, 0, 1);
+       if (r < 0) {
+          return r;
+       }
+
+       return ONE_BYTE[0] & 0xff;
+    }
 
-        ByteBuf take;
-        try {
-            take = isList.take();
-            boolean isReadable = take.isReadable();
-            if (checkEndOfInputOrError(take)) {
-                take.release();
-                return -1;
-            }
+    @Override
+    public synchronized void close() {
 
-            if (isReadable) {
-                return take.readInt();
-            } else {
-                take.release(); //We don't need `0`
-            }
+        releaseByteBuf();
 
-            return 0;
-        } catch (InterruptedException e) {
-            throw new IOException("Interrupted.", e);
-        }
+        cleanup(true);
     }
 
-    @Override
-    public void close() throws IOException {
-        if (isList != null) {
-            while (!isList.isEmpty()) {
-                try {
-                    isList.take().release();
-                } catch (InterruptedException e) {
-                    throw new IOException("Interrupted. Potential ByteBuf Leak.", e);
-                }
-            }
+    private synchronized void releaseByteBuf() {
+        if (current != null) {
+            current.release();
         }
-        super.close();
+
+        current = null;
+        buffer = null;
+    }
+
+    protected synchronized ByteBuffer awaitNext() {
+       while (isList.isEmpty()) {
+          if (end) {
+             return null;
+          }
+
+          try {
+             reading = true;
+             wait();
+             reading = false;
+          } catch (InterruptedException ie) {
+             // waiting uninterruptibly
+          }
+       }
+
+       current = isList.poll();
+       return current.nioBuffer().asReadOnlyBuffer();
+    }
+
+    public synchronized void complete(Throwable cause) {
+       this.cause = cause;
+       cleanup(cause != null);
+    }
+
+    protected void cleanup(boolean drain) {
+       if (drain) {
+          while (!isList.isEmpty()) {
+             isList.poll().release();
+          }
+       }
+
+       end = true;
+
+       if (reading) {
+          notifyAll();
+       }
     }
 
     @Override
     public int available() throws IOException {
-        ByteBuf peek = isList.peek();
-        if (peek != null && peek.isReadable()) {
-            return peek.readableBytes();
-        }
-        return 0;
+        return buffer == null ? 0 : buffer.remaining();
```

Comment:
should be synchronized

Reply:
ok

Reply:
Totally disagree. `NettyInputStream` has been, and remains, not thread safe. `read` and `close` methods have been, and are not `synchronized`, therefore it has always been assumed to be used by a single thread, or access is synchronized by out of band means.

Reply:
I see your point. We need to synchronize `isList` and `reading`. So only `awaitNext()`, `publish()`, and `cleanup()` are needed to be synchronized. We can remove it from `available`, `close`, `releaseByteBuf` and `complete`. Or do I miss something?

Reply:
No, just do not add to `available`.

`publish`, `awaitNext` and `close` interact with each other. A simple way to achieve total order of state transitions between things that are mostly done by two threads, and very rarely, is to use `synchronized`.

So we have one thread (Netty event loop) that receives byte buffers from the socket, and hands these buffers off to a different thread (Jersey thread that wants to `readEntity`). This is done through `publish`, and is potentially observed by `awaitNext`. `publish` also needs to see if maybe `close` has been called, so it does not assume anyone is going to release the buffers in the future. That's the purpose of `synchronized` around those three.

`synchronized` on `cleanup` is not harmful, but not necessary, because it is invoked from `synchronized` sections. `synchronized` on `available` I think is low cost, but because it does not achieve any improved correctness, I would recommend not adding it.

Reply:
Sounds like we agree here.
- Either we have `publish`, `awaitNext`, `complete` and `close` synchronized
- Or we have `publish`, `awaitNext`, and we move the synchronized from `complete` and `close` down to `cleanup`, where `end=true` and mainly `if (reading)` is used. 
Both are fine, though having `close()` synchronized could be a bit misleading as the NettyInputStream is not threadsafe.

Reply:
OK, as per discussion, I've implemented the B) way - compete and close (including available and releaseByteBuf) are not synchronized, cleanup became synchronized.

Reply:
LGTM